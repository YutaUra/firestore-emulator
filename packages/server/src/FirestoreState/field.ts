import { Value as v1Value } from "@firestore-emulator/proto/dist/google/firestore/v1/document";

import { NullValue } from "@firestore-emulator/proto/dist/google/protobuf/struct";

export type ValueObjectType = ReturnType<typeof v1Value.prototype.toObject>;
export interface FirestoreStateDocumentBaseField {
  toJSON(): { type: string; value: unknown };
  toV1ValueObject(): ValueObjectType;
  eq(other: FirestoreStateDocumentFields): boolean;
  lt(other: FirestoreStateDocumentFields): boolean;
  lte(other: FirestoreStateDocumentFields): boolean;
  gt(other: FirestoreStateDocumentFields): boolean;
  gte(other: FirestoreStateDocumentFields): boolean;
}

export class FirestoreStateDocumentStringField
  implements FirestoreStateDocumentBaseField
{
  type = "string_value" as const;
  constructor(readonly value: string) {}

  toJSON() {
    return { type: this.type, value: this.value };
  }

  toV1ValueObject(): ValueObjectType {
    return { string_value: this.value };
  }

  eq(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentStringField &&
      this.value === other.value
    );
  }

  lt(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentStringField &&
      this.value < other.value
    );
  }

  lte(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentStringField &&
      this.value <= other.value
    );
  }

  gt(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentStringField &&
      this.value > other.value
    );
  }

  gte(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentStringField &&
      this.value >= other.value
    );
  }
}

export class FirestoreStateDocumentNullField
  implements FirestoreStateDocumentBaseField
{
  type = "null_value" as const;
  value = null;
  toJSON() {
    return { type: this.type, value: null } as const;
  }

  toV1ValueObject(): ValueObjectType {
    return { null_value: NullValue.NULL_VALUE };
  }

  eq(other: FirestoreStateDocumentFields): boolean {
    return other.type === this.type;
  }

  lt(_other: FirestoreStateDocumentFields): boolean {
    return false;
  }

  lte(_other: FirestoreStateDocumentFields): boolean {
    return false;
  }

  gt(_other: FirestoreStateDocumentFields): boolean {
    return false;
  }

  gte(_other: FirestoreStateDocumentFields): boolean {
    return false;
  }
}

export class FirestoreStateDocumentBooleanField
  implements FirestoreStateDocumentBaseField
{
  type = "boolean_value" as const;
  constructor(readonly value: boolean) {}

  toJSON() {
    return { type: this.type, value: this.value };
  }

  toV1ValueObject(): ValueObjectType {
    return { boolean_value: this.value };
  }

  eq(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentBooleanField &&
      this.value === other.value
    );
  }

  lt(_other: FirestoreStateDocumentFields): boolean {
    return false;
  }

  lte(_other: FirestoreStateDocumentFields): boolean {
    return false;
  }

  gt(_other: FirestoreStateDocumentFields): boolean {
    return false;
  }

  gte(_other: FirestoreStateDocumentFields): boolean {
    return false;
  }
}

export class FirestoreStateDocumentIntegerField
  implements FirestoreStateDocumentBaseField
{
  type = "integer_value" as const;
  constructor(readonly value: number) {
    if (!Number.isInteger(value)) {
      throw new Error(`value must be integer. value=${value}`);
    }
  }

  toJSON() {
    return { type: this.type, value: this.value };
  }

  toV1ValueObject(): ValueObjectType {
    return { integer_value: this.value };
  }

  eq(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentIntegerField &&
      this.value === other.value
    );
  }

  lt(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentIntegerField &&
      this.value < other.value
    );
  }

  lte(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentIntegerField &&
      this.value <= other.value
    );
  }

  gt(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentIntegerField &&
      this.value > other.value
    );
  }

  gte(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentIntegerField &&
      this.value >= other.value
    );
  }
}

export class FirestoreStateDocumentDoubleField
  implements FirestoreStateDocumentBaseField
{
  type = "double_value" as const;
  constructor(readonly value: number) {}

  toJSON() {
    return { type: this.type, value: this.value };
  }

  toV1ValueObject(): ValueObjectType {
    return { double_value: this.value };
  }

  eq(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentDoubleField &&
      this.value === other.value
    );
  }

  lt(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentDoubleField &&
      this.value < other.value
    );
  }

  lte(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentDoubleField &&
      this.value <= other.value
    );
  }

  gt(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentDoubleField &&
      this.value > other.value
    );
  }

  gte(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentDoubleField &&
      this.value >= other.value
    );
  }
}

export class FirestoreStateDocumentTimestampField
  implements FirestoreStateDocumentBaseField
{
  type = "timestamp_value" as const;
  constructor(readonly value: { seconds: number; nanos: number }) {}

  static fromDate(date: Date) {
    return new FirestoreStateDocumentTimestampField({
      seconds: Math.floor(date.getTime() / 1000),
      nanos: (date.getTime() % 1000) * 1000000,
    });
  }

  toJSON() {
    return { type: this.type, value: this.value } as const;
  }

  toV1ValueObject(): ValueObjectType {
    return { timestamp_value: this.value };
  }

  eq(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentTimestampField &&
      this.value.seconds === other.value.seconds &&
      this.value.nanos === other.value.nanos
    );
  }

  lt(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentTimestampField &&
      (this.value.seconds < other.value.seconds ||
        (this.value.seconds === other.value.seconds &&
          this.value.nanos < other.value.nanos))
    );
  }

  lte(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentTimestampField &&
      (this.value.seconds < other.value.seconds ||
        (this.value.seconds === other.value.seconds &&
          this.value.nanos <= other.value.nanos))
    );
  }

  gt(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentTimestampField &&
      (this.value.seconds > other.value.seconds ||
        (this.value.seconds === other.value.seconds &&
          this.value.nanos > other.value.nanos))
    );
  }

  gte(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentTimestampField &&
      (this.value.seconds > other.value.seconds ||
        (this.value.seconds === other.value.seconds &&
          this.value.nanos >= other.value.nanos))
    );
  }
}

export class FirestoreStateDocumentBytesField
  implements FirestoreStateDocumentBaseField
{
  type = "bytes_value" as const;
  constructor(readonly value: Uint8Array) {}

  toJSON() {
    return { type: this.type, value: this.value } as const;
  }

  toV1ValueObject(): ValueObjectType {
    return { bytes_value: this.value };
  }

  eq(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentBytesField &&
      this.value.toString() === other.value.toString()
    );
  }

  lt(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentBytesField &&
      this.value.toString() < other.value.toString()
    );
  }

  lte(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentBytesField &&
      this.value.toString() <= other.value.toString()
    );
  }

  gt(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentBytesField &&
      this.value.toString() > other.value.toString()
    );
  }

  gte(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentBytesField &&
      this.value.toString() >= other.value.toString()
    );
  }
}

export class FirestoreStateDocumentReferenceField
  implements FirestoreStateDocumentBaseField
{
  type = "reference_value" as const;
  constructor(readonly value: string) {}

  toJSON() {
    return { type: this.type, value: this.value };
  }

  toV1ValueObject(): ValueObjectType {
    return { reference_value: this.value };
  }

  eq(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentReferenceField &&
      this.value === other.value
    );
  }

  lt(_other: FirestoreStateDocumentFields): boolean {
    return false;
  }

  lte(_other: FirestoreStateDocumentFields): boolean {
    return false;
  }

  gt(_other: FirestoreStateDocumentFields): boolean {
    return false;
  }

  gte(_other: FirestoreStateDocumentFields): boolean {
    return false;
  }
}

export class FirestoreStateDocumentGeoPointField
  implements FirestoreStateDocumentBaseField
{
  type = "geo_point_value" as const;
  constructor(readonly value: { latitude: number; longitude: number }) {}

  toJSON() {
    return { type: this.type, value: this.value };
  }

  toV1ValueObject(): ValueObjectType {
    return { geo_point_value: this.value };
  }

  eq(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentGeoPointField &&
      this.value.latitude === other.value.latitude &&
      this.value.longitude === other.value.longitude
    );
  }

  lt(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentGeoPointField &&
      (this.value.latitude < other.value.latitude ||
        this.value.longitude < other.value.longitude)
    );
  }

  lte(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentGeoPointField &&
      (this.value.latitude <= other.value.latitude ||
        this.value.longitude <= other.value.longitude)
    );
  }

  gt(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentGeoPointField &&
      (this.value.latitude > other.value.latitude ||
        this.value.longitude > other.value.longitude)
    );
  }

  gte(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentGeoPointField &&
      (this.value.latitude >= other.value.latitude ||
        this.value.longitude >= other.value.longitude)
    );
  }
}

export class FirestoreStateDocumentArrayField
  implements FirestoreStateDocumentBaseField
{
  type = "array_value" as const;
  constructor(readonly value: FirestoreStateDocumentFields[]) {}

  toJSON(): { type: string; value: unknown } {
    return {
      type: this.type,
      value: this.value.map((v) => v.toJSON()),
    };
  }

  toV1ValueObject(): ValueObjectType {
    return {
      array_value: { values: this.value.map((v) => v.toV1ValueObject()) },
    };
  }

  eq(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentArrayField &&
      this.value.length === other.value.length &&
      this.value.every((v, i) => {
        const item = other.value[i];
        if (!item) return false;
        return item.eq(v);
      })
    );
  }

  lt(_other: FirestoreStateDocumentFields): boolean {
    return false;
  }

  lte(_other: FirestoreStateDocumentFields): boolean {
    return false;
  }

  gt(_other: FirestoreStateDocumentFields): boolean {
    return false;
  }

  gte(_other: FirestoreStateDocumentFields): boolean {
    return false;
  }
}

export class FirestoreStateDocumentMapField
  implements FirestoreStateDocumentBaseField
{
  type = "map_value" as const;
  constructor(readonly value: Record<string, FirestoreStateDocumentFields>) {}

  toJSON(): { type: string; value: unknown } {
    return {
      type: this.type,
      value: Object.fromEntries(
        Object.entries(this.value).map(([k, v]) => [k, v.toJSON()])
      ),
    };
  }

  toV1ValueObject(): ValueObjectType {
    return {
      map_value: {
        fields: Object.fromEntries(
          Object.entries(this.value).map(([k, v]) => [k, v.toV1ValueObject()])
        ),
      },
    };
  }

  eq(other: FirestoreStateDocumentFields): boolean {
    return (
      other instanceof FirestoreStateDocumentMapField &&
      Object.keys(this.value).length === Object.keys(other.value).length &&
      Object.entries(this.value).every(([k, v]) => {
        const item = other.value[k];
        if (!item) return false;
        return item.eq(v);
      })
    );
  }

  lt(_other: FirestoreStateDocumentFields): boolean {
    return false;
  }

  lte(_other: FirestoreStateDocumentFields): boolean {
    return false;
  }

  gt(_other: FirestoreStateDocumentFields): boolean {
    return false;
  }

  gte(_other: FirestoreStateDocumentFields): boolean {
    return false;
  }
}

export type FirestoreStateDocumentFields =
  | FirestoreStateDocumentStringField
  | FirestoreStateDocumentNullField
  | FirestoreStateDocumentBooleanField
  | FirestoreStateDocumentIntegerField
  | FirestoreStateDocumentDoubleField
  | FirestoreStateDocumentTimestampField
  | FirestoreStateDocumentBytesField
  | FirestoreStateDocumentReferenceField
  | FirestoreStateDocumentGeoPointField
  | FirestoreStateDocumentArrayField
  | FirestoreStateDocumentMapField;

export const convertV1DocumentField = (
  field: v1Value
): FirestoreStateDocumentFields => {
  if (field.has_string_value)
    return new FirestoreStateDocumentStringField(field.string_value);
  if (field.has_null_value) return new FirestoreStateDocumentNullField();
  if (field.has_boolean_value)
    return new FirestoreStateDocumentBooleanField(field.boolean_value);
  if (field.has_integer_value)
    return new FirestoreStateDocumentIntegerField(field.integer_value);
  if (field.has_double_value)
    return new FirestoreStateDocumentDoubleField(field.double_value);
  if (field.has_timestamp_value)
    return new FirestoreStateDocumentTimestampField({
      nanos: field.timestamp_value.nanos ?? 0,
      seconds: field.timestamp_value.seconds ?? 0,
    });
  if (field.has_bytes_value)
    return new FirestoreStateDocumentBytesField(field.bytes_value);
  if (field.has_reference_value)
    return new FirestoreStateDocumentReferenceField(field.reference_value);
  if (field.has_geo_point_value)
    return new FirestoreStateDocumentGeoPointField({
      latitude: field.geo_point_value.latitude ?? 0,
      longitude: field.geo_point_value.longitude ?? 0,
    });
  if (field.has_array_value)
    return new FirestoreStateDocumentArrayField(
      (field.array_value.values ?? []).map(convertV1DocumentField)
    );
  if (field.has_map_value)
    return new FirestoreStateDocumentMapField(
      Object.fromEntries(
        Array.from(field.map_value.fields.entries() ?? []).map(([k, v]) => [
          k,
          convertV1DocumentField(v),
        ])
      )
    );
  throw new Error(`unknown field type. field=${JSON.stringify(field)}`);
};

export const convertV1Value = (
  value: v1Value
): ReturnType<typeof v1Value.prototype.toObject> => {
  return convertV1DocumentField(value).toV1ValueObject();
};
