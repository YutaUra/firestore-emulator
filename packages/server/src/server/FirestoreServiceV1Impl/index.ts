import type { Document } from "@firestore-emulator/proto/dist/google/firestore/v1/document";
import type {
  BatchGetDocumentsRequest,
  BatchWriteRequest,
  BatchWriteResponse,
  BeginTransactionRequest,
  CommitRequest,
  CreateDocumentRequest,
  DeleteDocumentRequest,
  GetDocumentRequest,
  ListCollectionIdsRequest,
  ListCollectionIdsResponse,
  ListDocumentsRequest,
  ListDocumentsResponse,
  ListenResponse,
  PartitionQueryRequest,
  PartitionQueryResponse,
  RollbackRequest,
  RunAggregationQueryRequest,
  RunQueryRequest,
  UpdateDocumentRequest,
  WriteRequest,
  WriteResponse,
} from "@firestore-emulator/proto/dist/google/firestore/v1/firestore";
import {
  BatchGetDocumentsResponse,
  BeginTransactionResponse,
  CommitResponse,
  ListenRequest,
  RunAggregationQueryResponse,
  RunQueryResponse,
  UnimplementedFirestoreService,
} from "@firestore-emulator/proto/dist/google/firestore/v1/firestore";
import { Empty } from "@firestore-emulator/proto/dist/google/protobuf/empty";
import type {
  ServerDuplexStream,
  ServerUnaryCall,
  ServerWritableStream,
  sendUnaryData,
  UntypedHandleCall,
} from "@grpc/grpc-js";
import { FirestoreEmulatorError } from "../../error/error";
import type { FirestoreState } from "../../FirestoreState";
import { TimestampFromDate, TimestampFromNow } from "../../FirestoreState";
import {
  FirestoreStateDocumentDoubleField,
  FirestoreStateDocumentIntegerField,
  FirestoreStateDocumentNullField,
  type ValueObjectType,
} from "../../FirestoreState/field";

export class FirestoreServiceV1Impl extends UnimplementedFirestoreService {
  #state: FirestoreState;
  constructor(state: FirestoreState) {
    super();

    this.#state = state;
  }
  override GetDocument(
    _call: ServerUnaryCall<GetDocumentRequest, Document>,
    _callback: sendUnaryData<Document>,
  ): void {
    console.error("Method<GetDocument> not implemented.");
    throw new Error("Method<GetDocument> not implemented.");
  }
  override ListDocuments(
    _call: ServerUnaryCall<ListDocumentsRequest, ListDocumentsResponse>,
    _callback: sendUnaryData<ListDocumentsResponse>,
  ): void {
    console.error("Method<ListDocuments> not implemented.");
    throw new Error("Method<ListDocuments> not implemented.");
  }
  override UpdateDocument(
    _call: ServerUnaryCall<UpdateDocumentRequest, Document>,
    _callback: sendUnaryData<Document>,
  ): void {
    console.error("Method<UpdateDocument> not implemented.");
    throw new Error("Method<UpdateDocument> not implemented.");
  }
  override DeleteDocument(
    _call: ServerUnaryCall<DeleteDocumentRequest, Empty>,
    _callback: sendUnaryData<Empty>,
  ): void {
    console.error("Method<DeleteDocument> not implemented.");
    throw new Error("Method<DeleteDocument> not implemented.");
  }
  override BatchGetDocuments(
    call: ServerWritableStream<
      BatchGetDocumentsRequest,
      BatchGetDocumentsResponse
    >,
  ): void {
    const date = TimestampFromNow();
    const tx = call.request.has_new_transaction
      ? Uint8Array.from([17, 2, 0, 0, 0, 0, 0, 0, 0])
      : null;
    if (tx) {
      call.write(
        BatchGetDocumentsResponse.fromObject({
          transaction: tx,
        }),
      );
    }
    for (const documentPath of call.request.documents) {
      const document = this.#state.getDocument(documentPath);
      call.write(
        BatchGetDocumentsResponse.fromObject({
          found: document.metadata.hasExist
            ? document.toV1DocumentObject()
            : undefined,
          missing: document.metadata.hasExist ? undefined : documentPath,
          read_time: date,
        }),
      );
    }

    call.end();
  }
  override BeginTransaction(
    _call: ServerUnaryCall<BeginTransactionRequest, BeginTransactionResponse>,
    callback: sendUnaryData<BeginTransactionResponse>,
  ): void {
    callback(
      null,
      BeginTransactionResponse.fromObject({
        // dummy transaction id
        transaction: Uint8Array.from([17, 2, 0, 0, 0, 0, 0, 0, 0]),
      }),
    );
  }

  override Commit(
    call: ServerUnaryCall<CommitRequest, CommitResponse>,
    callback: sendUnaryData<CommitResponse>,
  ): void {
    const date = new Date();
    try {
      const results = call.request.writes.map((write) => {
        return this.#state.writeV1Document(date, write);
      });
      callback(
        null,
        CommitResponse.fromObject({
          commit_time: TimestampFromDate(date),
          write_results: results.map((result) => result.toObject()),
        }),
      );
    } catch (err) {
      if (err instanceof FirestoreEmulatorError) {
        callback(
          {
            cause: err,
            code: err.code,
            message: err.message,
            name: "Error",
          },
          null,
        );
        return;
      }
      console.error(err);
      callback(
        {
          cause: err,
          message: "Something went wrong",
          name: "Error",
        },
        null,
      );
    }
  }
  override Rollback(
    _call: ServerUnaryCall<RollbackRequest, Empty>,
    callback: sendUnaryData<Empty>,
  ): void {
    callback(null, Empty.fromObject({}));
  }
  override RunQuery(
    call: ServerWritableStream<RunQueryRequest, RunQueryResponse>,
  ): void {
    const date = new Date();
    const results = this.#state.v1Query(
      call.request.parent,
      call.request.structured_query,
    );

    let transaction: Uint8Array | undefined = call.request.transaction;
    if (call.request.consistency_selector === "new_transaction") {
      transaction = crypto.getRandomValues(new Uint8Array(12));
    }

    if (results.length > 0) {
      results.forEach((result, i, arr) => {
        call.write(
          RunQueryResponse.fromObject({
            document: result.toV1DocumentObject(),
            done: i === arr.length - 1,
            read_time: TimestampFromDate(date),
            skipped_results: 0,
            transaction,
          }),
        );
      });
    } else {
      call.write(
        RunQueryResponse.fromObject({
          done: true,
          read_time: TimestampFromDate(date),
          skipped_results: 0,
          transaction,
        }),
      );
    }

    call.end();
  }
  override RunAggregationQuery(
    call: ServerWritableStream<
      RunAggregationQueryRequest,
      RunAggregationQueryResponse
    >,
  ): void {
    const date = TimestampFromNow();
    const results = this.#state.v1Query(
      call.request.parent,
      call.request.structured_aggregation_query.structured_query,
    );
    let transaction: Uint8Array | undefined = call.request.transaction;
    if (call.request.consistency_selector === "new_transaction") {
      transaction = crypto.getRandomValues(new Uint8Array(12));
    }
    if (call.request.has_explain_options) {
      const error = new Error(
        "Explain for aggregation query is not supported.",
      );
      console.error(error);
      call.emit("error");
      throw error;
    }

    const aggregateResult = Object.fromEntries(
      call.request.structured_aggregation_query.aggregations.map<
        [string, ValueObjectType]
      >((agg) => {
        if (agg.has_count)
          return [agg.alias, { integer_value: results.length }];
        if (agg.has_sum) {
          const sum = results.reduce<
            | FirestoreStateDocumentIntegerField
            | FirestoreStateDocumentDoubleField
          >((acc, cur) => {
            const field = cur.getField(agg.sum.field.field_path);
            if (!field) return acc;
            if (field.type === "integer_value") return acc.add(field);
            if (field.type === "double_value") return field.add(acc);
            return acc;
          }, new FirestoreStateDocumentIntegerField(0));
          return [agg.alias, sum.toV1ValueObject()];
        }
        if (agg.has_avg) {
          const sum = results.reduce<
            | FirestoreStateDocumentIntegerField
            | FirestoreStateDocumentDoubleField
          >((acc, cur) => {
            const field = cur.getField(agg.avg.field.field_path);
            if (!field) return acc;
            if (field.type === "integer_value") return acc.add(field);
            if (field.type === "double_value") return field.add(acc);
            return acc;
          }, new FirestoreStateDocumentIntegerField(0));
          return [
            agg.alias,
            new FirestoreStateDocumentDoubleField(
              sum.value / results.length,
            ).toV1ValueObject(),
          ];
        }
        return [
          agg.alias,
          new FirestoreStateDocumentNullField().toV1ValueObject(),
        ];
      }),
    );
    const res = RunAggregationQueryResponse.fromObject({
      result: {
        aggregate_fields: aggregateResult,
      },
      read_time: date,
      transaction,
    });
    call.write(res);
    call.end();
  }
  override PartitionQuery(
    _call: ServerUnaryCall<PartitionQueryRequest, PartitionQueryResponse>,
    _callback: sendUnaryData<PartitionQueryResponse>,
  ): void {
    console.error("Method<PartitionQuery> not implemented.");
    throw new Error("Method<PartitionQuery> not implemented.");
  }
  override Write(_call: ServerDuplexStream<WriteRequest, WriteResponse>): void {
    console.error("Method<Write> not implemented.");
    throw new Error("Method<Write> not implemented.");
  }
  override Listen(
    call: ServerDuplexStream<ListenRequest, ListenResponse>,
  ): void {
    call.once("data", (request) => {
      if (!(request instanceof ListenRequest)) {
        console.error("Invalid request type");
        call.end();
        throw new Error("Invalid request type");
      }
      this.#state.v1Listen(
        request,
        (value) => {
          call.write(value);
        },
        (handler) => {
          call.once("end", handler);
        },
      );
    });
  }
  override ListCollectionIds(
    _call: ServerUnaryCall<ListCollectionIdsRequest, ListCollectionIdsResponse>,
    _callback: sendUnaryData<ListCollectionIdsResponse>,
  ): void {
    console.error("Method<ListCollectionIds> not implemented.");
    throw new Error("Method<ListCollectionIds> not implemented.");
  }
  override BatchWrite(
    _call: ServerUnaryCall<BatchWriteRequest, BatchWriteResponse>,
    _callback: sendUnaryData<BatchWriteResponse>,
  ): void {
    console.error("Method<BatchWrite> not implemented.");
    throw new Error("Method<BatchWrite> not implemented.");
  }
  override CreateDocument(
    _call: ServerUnaryCall<CreateDocumentRequest, Document>,
    _callback: sendUnaryData<Document>,
  ): void {
    console.error("Method<CreateDocument> not implemented.");
    throw new Error("Method<CreateDocument> not implemented.");
  }

  [name: string]: UntypedHandleCall;
}
