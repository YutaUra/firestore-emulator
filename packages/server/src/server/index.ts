import { Server, ServerCredentials } from '@grpc/grpc-js'

import { FirestoreState } from '../FirestoreState'

import { FirestoreServiceV1Impl } from './FirestoreServiceV1Impl'

export class FirestoreServer {
  state: FirestoreState
  private readonly server: Server
  constructor() {
    this.server = new Server()

    this.state = new FirestoreState()
    this.server.addService(
      FirestoreServiceV1Impl.definition,
      new FirestoreServiceV1Impl(this.state),
    )
  }

  async start(port: number): Promise<void> {
    return new Promise((resolve, reject) => {
      this.server.bindAsync(
        `0.0.0.0:${port}`,
        ServerCredentials.createInsecure(),
        (err) => {
          if (err) {
            reject(err)
            return
          }
          resolve(undefined)
        },
      )
    })
  }

  stop() {
    this.server.forceShutdown()
  }
}
