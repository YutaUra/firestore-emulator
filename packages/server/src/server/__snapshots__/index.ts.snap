// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`can create document 1`] = `
{
  "projects": {
    "mealmatesnetwork-dev0": {
      "databases": {
        "(default)": {
          "collections": {
            "test": {
              "documents": {
                "test": {
                  "collections": {},
                  "fields": {
                    "foo": "bar",
                  },
                  "path": "projects/mealmatesnetwork-dev0/databases/(default)/test/test",
                },
              },
              "path": "projects/mealmatesnetwork-dev0/databases/(default)/test",
            },
          },
          "path": "projects/mealmatesnetwork-dev0/databases/(default)",
        },
      },
      "path": "projects/mealmatesnetwork-dev0",
    },
  },
}
`;

exports[`can delete document 1`] = `
{
  "projects": {
    "mealmatesnetwork-dev0": {
      "databases": {
        "(default)": {
          "collections": {
            "test": {
              "documents": {
                "test": {
                  "collections": {},
                  "fields": {},
                  "path": "projects/mealmatesnetwork-dev0/databases/(default)/test/test",
                },
              },
              "path": "projects/mealmatesnetwork-dev0/databases/(default)/test",
            },
          },
          "path": "projects/mealmatesnetwork-dev0/databases/(default)",
        },
      },
      "path": "projects/mealmatesnetwork-dev0",
    },
  },
}
`;

exports[`can get document 1`] = `
{
  "projects": {
    "mealmatesnetwork-dev0": {
      "databases": {
        "(default)": {
          "collections": {
            "test": {
              "documents": {
                "test": {
                  "collections": {},
                  "fields": {},
                  "path": "projects/mealmatesnetwork-dev0/databases/(default)/test/test",
                },
              },
              "path": "projects/mealmatesnetwork-dev0/databases/(default)/test",
            },
          },
          "path": "projects/mealmatesnetwork-dev0/databases/(default)",
        },
      },
      "path": "projects/mealmatesnetwork-dev0",
    },
  },
}
`;

exports[`can list documents 1`] = `
{
  "projects": {
    "mealmatesnetwork-dev0": {
      "databases": {
        "(default)": {
          "collections": {
            "test": {
              "documents": {
                "test": {
                  "collections": {},
                  "fields": {
                    "foo": "bar1",
                  },
                  "path": "projects/mealmatesnetwork-dev0/databases/(default)/test/test",
                },
                "test2": {
                  "collections": {},
                  "fields": {
                    "foo": "bar2",
                  },
                  "path": "projects/mealmatesnetwork-dev0/databases/(default)/test/test2",
                },
                "test3": {
                  "collections": {},
                  "fields": {
                    "foo": "bar3",
                  },
                  "path": "projects/mealmatesnetwork-dev0/databases/(default)/test/test3",
                },
              },
              "path": "projects/mealmatesnetwork-dev0/databases/(default)/test",
            },
          },
          "path": "projects/mealmatesnetwork-dev0/databases/(default)",
        },
      },
      "path": "projects/mealmatesnetwork-dev0",
    },
  },
}
`;

exports[`can list documents with where 1`] = `
{
  "projects": {
    "mealmatesnetwork-dev0": {
      "databases": {
        "(default)": {
          "collections": {
            "test": {
              "documents": {
                "test": {
                  "collections": {},
                  "fields": {
                    "foo": "bar1",
                  },
                  "path": "projects/mealmatesnetwork-dev0/databases/(default)/test/test",
                },
                "test2": {
                  "collections": {},
                  "fields": {
                    "foo": "bar2",
                  },
                  "path": "projects/mealmatesnetwork-dev0/databases/(default)/test/test2",
                },
                "test3": {
                  "collections": {},
                  "fields": {
                    "foo": "bar3",
                  },
                  "path": "projects/mealmatesnetwork-dev0/databases/(default)/test/test3",
                },
              },
              "path": "projects/mealmatesnetwork-dev0/databases/(default)/test",
            },
          },
          "path": "projects/mealmatesnetwork-dev0/databases/(default)",
        },
      },
      "path": "projects/mealmatesnetwork-dev0",
    },
  },
}
`;

exports[`can set document 1`] = `
{
  "projects": {
    "mealmatesnetwork-dev0": {
      "databases": {
        "(default)": {
          "collections": {
            "test": {
              "documents": {
                "test": {
                  "collections": {},
                  "fields": {
                    "foo": "bar",
                  },
                  "path": "projects/mealmatesnetwork-dev0/databases/(default)/test/test",
                },
              },
              "path": "projects/mealmatesnetwork-dev0/databases/(default)/test",
            },
          },
          "path": "projects/mealmatesnetwork-dev0/databases/(default)",
        },
      },
      "path": "projects/mealmatesnetwork-dev0",
    },
  },
}
`;

exports[`can set document with merge 1`] = `
{
  "projects": {
    "mealmatesnetwork-dev0": {
      "databases": {
        "(default)": {
          "collections": {
            "test": {
              "documents": {
                "test": {
                  "collections": {},
                  "fields": {
                    "bar": "baz",
                    "foo": "bar",
                  },
                  "path": "projects/mealmatesnetwork-dev0/databases/(default)/test/test",
                },
              },
              "path": "projects/mealmatesnetwork-dev0/databases/(default)/test",
            },
          },
          "path": "projects/mealmatesnetwork-dev0/databases/(default)",
        },
      },
      "path": "projects/mealmatesnetwork-dev0",
    },
  },
}
`;

exports[`can update document 1`] = `
{
  "projects": {
    "mealmatesnetwork-dev0": {
      "databases": {
        "(default)": {
          "collections": {
            "test": {
              "documents": {
                "test": {
                  "collections": {},
                  "fields": {
                    "foo": "baz",
                  },
                  "path": "projects/mealmatesnetwork-dev0/databases/(default)/test/test",
                },
              },
              "path": "projects/mealmatesnetwork-dev0/databases/(default)/test",
            },
          },
          "path": "projects/mealmatesnetwork-dev0/databases/(default)",
        },
      },
      "path": "projects/mealmatesnetwork-dev0",
    },
  },
}
`;

exports[`get document after create 1`] = `
{
  "projects": {
    "mealmatesnetwork-dev0": {
      "databases": {
        "(default)": {
          "collections": {
            "test": {
              "documents": {
                "test": {
                  "collections": {},
                  "fields": {
                    "foo": "bar",
                  },
                  "path": "projects/mealmatesnetwork-dev0/databases/(default)/test/test",
                },
              },
              "path": "projects/mealmatesnetwork-dev0/databases/(default)/test",
            },
          },
          "path": "projects/mealmatesnetwork-dev0/databases/(default)",
        },
      },
      "path": "projects/mealmatesnetwork-dev0",
    },
  },
}
`;
